### PR TITLE
Airbyte-ci: Add ClickPipelineContext and update test to use

### PR DESCRIPTION
**Pull request by bnchrch**
## Overview
1. Adds the `ClickPipelineContext` to give developers a primitive to merge click commands, dagger clients and pipeline contexts into one
2. Updates `airbyte-ci test` to use `ClickPipelineContext`
2. Adds new click decorators for automatically attaching click arguments to the click context object so that child commands may use them

### CODE REVIEW

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:237]: @alafanechere @erohmensing Added a new decorator to standardize appending to the context, and make it closer to argument/option decorators.

basically replaces
```
ctx.obj["is_local"] = value
```

Not super attached but felt useful

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:236]: @alafanechere @erohmensing This is what merges all click arguments/options into the context object by default.

basically another name would be `@pass_all_click_args_to_child`

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/playground/commands.py
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/playground/commands.py:9]: @alafanechere @erohmensing This is the new kid on the block for instantiating a pipeline/dagger client

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/playground/commands.py
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/playground/commands.py:26]: Will finish one the asyncclick PR is ready

File: airbyte-ci/connectors/pipelines/pipelines/models/singleton.py
Comment (cpdeethree) [inline airbyte-ci/connectors/pipelines/pipelines/models/singleton.py:4]: Suggest adding test cases here like I added in this PR:

https://github.com/airbytehq/aircmd/pull/16

notably these test cases surface a small issue with my implementation that is almost certainly not applicable to your use case - if you load a module using Singleton dynamically (such as in the plugin use case) it is theoretically possible to instantiate multiple copies of Singleton. I would take a look at the failing test case and just ensure it will never apply to your use case or push a fix.

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (cpdeethree) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:70]: Nice. I'm glad this moves us towards One Client in connectors land. One thing I learned with Prefect is that doing this means you cannot call this client multiple thread pools (which you should not be doing anyway). If it does happen, you'll know though, because it will throw an error. Just something to be aware of when using this pattern with other asynchronous libraries. I think few async libraries make cross thread pool calls, so you should run into no issues.

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (cpdeethree) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:19]: You can probably get rid of this if you aren't using Prefect. It only matters if you are wrapping function sin Flow calls

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:36]: Can we add the CI suffix to the printed message :D

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:245]: I saw this in use before I got down to my comment. This is sexy 🔥

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:36]: Done!

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:18]: I'd love a docstring to understand the purpose of this property.
I believe its content is static, should we add the params attribute it at init time?

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:33]: Why are we declaring a nested class here?

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:50]: Can we avoid the inline import? 
I would also prefer to use `anyio` primitives instead of `asyncio`

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:55]: Is the doubled `if not self._dagger_client` statement intended?

File: airbyte-ci/connectors/pipelines/pipelines/stolen/__init__.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/stolen/__init__.py:3]: Can't find anything under this `stolen` package.

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:12]: I'd love to see this class unit tested

**Review by alafanechere — CHANGES_REQUESTED** [review]
I love the direction! You're becoming a decorator wizard!
My main concerns are the following:
1. Some command options end up in the command function signature, some others don't. I'd love to have a strict command signature policy. I'd be fine if all the options ended up being merged in the context obj.

2. You wrote *Adds the ClickPipelineContext to give developers a primitive to merge click commands, dagger clients and pipeline contexts into one* .  I don't see how click context gets merged with the current `PipelieContext` class. So far the benefits I've grasped are:
-  less boilerplate code to load option values to the context.
- A singleton dagger client

Let me know if I missed something!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/commands.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/commands.py:196]: Yup! can do!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:13]: Im not sure I follow?

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:50]: Sure thing. Whats the reason for prefering anyio?

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:55]: Taken from Conors aircmd. I imagine its a double check that the client has not been initialized when the function was trying to get a lcok

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:33]: This class was lifted from Aircmd with minor modifications. Didnt want to mess with an implementation that was working

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:13]: Sorry, missings words.
Can we `from pipelines.models.contexts.click_pipeline_context import pass_pipeline_context`

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:57]: `anyio` is the async framework wrapper that dagger suggested that we use. It can run on top of `asyncio` or `trio`. And  in tests, `anyio` and `asyncio` fixture do not play well together.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (erohmensing) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:75]: @bnchrch With removing the `try/except` does this still make sense? Previously we were returning True if the container ran and handling exceptions. Now it looks like we will either return the container or hit an exception which will not be aborted

(currently trying to refactor my code which was built off of this and am confused here)

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:13]: +1, since pretty much everywhere will be using this

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:75]: Good call out! Yes this can go away!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (erohmensing) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:27]: I'm not sure if it matters or if it is easy to change. One thing that has been weird for me is the fact that we name both the `click.Context` and `ClickPipelineContext` `ctx` as params. The type hints help, but I still catch myself trying to use one's method for the other. 

In addition, there have been places where I want to use both (To be seen whether my use case for doing this is actually at all legit, but...) To accomplish this I have 
```
async def fix(ctx: click.Context, pipeline_ctx: ClickPipelineContext):
```
where fix is a click group... I did not think this would work at all and yet it did. Just some food for thought

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (erohmensing) [inline airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:173]: 😍

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:251]: ❓ In the case that setting the parameter is more complicated than just setting a value, would you suggest adding a util method and referencing it here? or doing it in the method itself?

I would potentially be looking into making this accept a coroutine in addition to a callable or value, if it fixes my problems with trying to append something. But that doesn't have to be done here :)

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:274]: Direct follow up to my question above about more complicated callables to set things 😄

File: airbyte-ci/connectors/pipelines/pipelines/cli/click_decorators.py
Comment (erohmensing) [inline airbyte-ci/connectors/pipelines/pipelines/cli/click_decorators.py:1]: I'll admit a lot of this stuff is pretty nitty gritty to understand at first glance. I don't think the code can be simplified much, but unit tests to show behavior might help!

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:29]: I think we add to the params when we add arguments to click commands? 

I've been pulling params from this property in the way that I used to access the other type of context, i.e. 

If ctx is click.Context: `ctx.obj.get("key")`
if ctx is ClickPipelineContext: `ctx.params.get("key")`

and I guess the idea is that these `params` contain both the click context object keys and the click command arguments

Would love confirmation :)

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:64]: I had the same question, would love to get conor in on this one so that we can add a comment if that's the case, cause IMO it's confusing and easy to try to "clean up" later

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:76]: @cpdeethree can you link any info you know about this? I'd be interested in learning more, i've been running into some messes trying to share the client during command chaining and potentially trying to run 4 different (read-only) checks in parallel

File: airbyte-ci/connectors/pipelines/tests/conftest.py
Comment (erohmensing) [inline airbyte-ci/connectors/pipelines/tests/conftest.py:27]: ❓ I think this makes sense and is fine, but don't see it used anywhere, just wondering why we split this

**Review by erohmensing — COMMENTED** [review]
I have a few more questions but this is looking great, and I'm using it already on my branch :)

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:27]: Hmm 🤔 

So I see what you mean on the naming.

Im not sure a good general path, but we could update the naming of the variable to be `pipeline_context` and just socially enforce it?

In the case of using the `ClickPipelineContext` and the `ClickContext`

You could/should use the internal ref to the `clickcontext` 
```py
async def fix(pipeline_ctx: ClickPipelineContext):
    click_context = pipeline_ctx._click_context
```

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:251]: Yup! Would definately recommend a named function instead of a lambda.

Also would welcome a coroutine version! Ill look into updating it to allow that

File: airbyte-ci/connectors/pipelines/pipelines/cli/click_decorators.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/cli/click_decorators.py:1]: Good call out, will add!

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:29]: That is the idea! Though. It doesnt have to be. But the idea was to always be merging the idea of parent arguments/options with current argument/options so the way you reference "branch_name" or "connector_name" is the same no matter what level of the CLI command hierarchy youre at

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:64]: Calling in the reinforcements! @cpdeethree, The idea is to ensure nothing got intialized when getting the lock right?

File: airbyte-ci/connectors/pipelines/tests/conftest.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/tests/conftest.py:27]: Ahh I can merge back together. There was an approach I was using to mocking out our `get_dagger_client` method in the `ClickPipelineCOntext` which needed the dagger connection as a fixture instead of the client.

I didnt end up going that route for the test in the end

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py:27]: > You could/should use the internal ref to the clickcontext
ooh, i'll try this 

> Im not sure a good general path, but we could update the naming of the variable to be pipeline_context and just socially enforce it? 
I think the below looks nice, I would be into it! 
```
async def fix(pipeline_ctx: ClickPipelineContext):
```

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:29]: It's not a blocker for me, would just appreciate a docstring 😄

**Comment by alafanechere** [comment]
@bnchrch let me know when you need a final review.
Let me know if you consider doing the following:
> Some command options end up in the command function signature, some others don't. I'd love to have a strict command signature policy. I'd be fine if all the options ended up being merged in the context obj.

I'd appreciate context on this too, but it's not blocking:

> You wrote *Adds the ClickPipelineContext to give developers a primitive to merge click commands, dagger clients and pipeline contexts into one* . I don't see how click context gets merged with the current PipelineContext class.

**Comment by bnchrch** [comment]
> Some command options end up in the command function signature, some others don't. I'd love to have a strict command signature policy. I'd be fine if all the options ended up being merged in the context obj.

@alafanechere I dont think this is achievable without extending / wrapping `@click.command` Which I think is a bit too big of a lift to restrict use

> You wrote Adds the ClickPipelineContext to give developers a primitive to merge click commands, dagger clients and pipeline contexts into one . I don't see how click context gets merged with the current PipelineContext class.

Ah, `test` may not be the best example but if you look here
https://github.com/airbytehq/airbyte/blob/24297dff8968c715e4c042de5e49800fabbadc9b/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py#L20
```py
@click.command()
@click.argument("poetry_package_path")
@click.option("--test-directory", default="tests", help="The directory containing the tests to run.")
@pass_pipeline_context
@click_ignore_unused_kwargs
async def test(
    pipeline_context: ClickPipelineContext,
    poetry_package_path: str,
    test_directory: str,
):
```

`pipeline_context` is the only context object necessary.
1. You can get the instantiated dagger client via `pipeline_context.get_dagger_client()`
2. You can get the click context via `pipeline_context._click_context`
3. You can access click params and object via `pipeline_context.params`

**Comment by alafanechere** [comment]
> > Some command options end up in the command function signature, some others don't. I'd love to have a strict command signature policy. I'd be fine if all the options ended up being merged in the context obj.
> 
> @alafanechere I dont think this is achievable without extending / wrapping `@click.command` Which I think is a bit too big of a lift to restrict use.

Even if we don't restrict use, can we make sure that we follow this policy? In other words, in this PR can you avoid passing the options to the signature and just pull their values from the context? Or is it something click forces you to do?


> > You wrote Adds the ClickPipelineContext to give developers a primitive to merge click commands, dagger clients and pipeline contexts into one . I don't see how click context gets merged with the current PipelineContext class.
> 
> Ah, `test` may not be the best example but if you look here
> 
> https://github.com/airbytehq/airbyte/blob/24297dff8968c715e4c042de5e49800fabbadc9b/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/test/commands.py#L20
> 
> ```python
> @click.command()
> @click.argument("poetry_package_path")
> @click.option("--test-directory", default="tests", help="The directory containing the tests to run.")
> @pass_pipeline_context
> @click_ignore_unused_kwargs
> async def test(
>     pipeline_context: ClickPipelineContext,
>     poetry_package_path: str,
>     test_directory: str,
> ):
> ```
> 
> `pipeline_context` is the only context object necessary.
> 
> 1. You can get the instantiated dagger client via `pipeline_context.get_dagger_client()`
> 2. You can get the click context via `pipeline_context._click_context`
> 3. You can access click params and object via `pipeline_context.params`

@bnchrch yep it's nice.
My question was rather "how easy is it now to avoid instantiating [ConnectorContext](https://github.com/airbytehq/airbyte/blob/8bfe2aa610a2121059b879876eed4ab3da2915b6/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py#L63) with such a bulky approach?"

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:71]: I'd love to have a public `PipelineContext.dagger_client` attribute / property.
It will be more convenient than running `await pipeline_context.get_dagger_client()`
Could we expose an async init class method to initialize PipelineContext where the `_dagger_client` attribute would be set? 
```python
pipeline_context = await PipelineContext.init(data)
pipeline_context.dagger_client
```
The benefit of `get_dagger_client` is that you can pass custom pipeline name though.
Let me know your opinion about this.

Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py:71]:
```suggestion
    async def get_dagger_client(self, pipeline_name: Optional[str] = None) -> Client:
```
I don't think this `client` parameter is used in your function.

**Review by alafanechere — APPROVED** [review]
Nothing blocking! Thank you for the effort! 
Please run a pre-release + connector test manually before merging 🙏

File: airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/models/contexts/click_pipeline_context.py]: Closing the loop:

I left a reply in slack on this but TLDR I dont think the gains of avoiding the `async` call here are worth it.


The Reply:
> So If we cant await in the constructor
> 
> Then we would have to explicitly call
> `pipeline_context = await PipelineContext.init(data)
> `
> somewhere in the command hierarchy
> 
> which has a few downsides (as far as im understanding it)
> 
> - extra boilerplate
> - an extra thing to remember to run
> - that extra thing may be called well before the area of the code your working on (i.e. in the group and your working on a subcommand) leaving you unaware of its existence or necessity 
> - and an extra piece to consider when chaining together subcommands
> 
> The upside is we dont use await
> 
> Im not sure thats worth the speed trade off
> 
> Though Im not sure the speed price were paying

**Comment by bnchrch** [comment]
- [ ] Tests are passing
- [ ] Connector Prerelease was successful
- [ ] Connector test was successful

**Pull request merged**
